test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildServerIntegrationTest.cs

```diff
@@ -143,28 +144,55 @@ public async Task Build_ServerConnectionMutexCreationFails_FallsBackToInProcessR
         {
-            var toolAssembly = typeof(Application).Assembly.Location;
-            var result = await DotnetMSBuild(
-                "Build",
-                $"/p:_RazorForceBuildServer=true /p:_RazorToolAssembly={toolAssembly}",
-                suppressBuildServer: true); // We don't want to specify a pipe name
+            // Publish rzc tool to a temporary path
+            var solutionRoot = TestPathUtilities.GetSolutionRootDirectory("Razor");
+            var toolAssemblyDirectory = Path.Combine(solutionRoot, "src", "Microsoft.AspNetCore.Razor.Tools");
+            var toolAssemblyPath = Path.Combine(toolAssemblyDirectory, "Microsoft.AspNetCore.Razor.Tools.csproj");
+            var projectDirectory = new TestProjectDirectory(solutionRoot, toolAssemblyDirectory, toolAssemblyPath);
+            var publishDir = Path.Combine(Path.GetTempPath(), "Razor", Path.GetRandomFileName(), "RzcPublish");
+            var publishResult = await MSBuildProcessManager.RunProcessAsync(projectDirectory, $"/t:Publish /p:PublishDir=\"{publishDir}\"");
 
-            Assert.BuildPassed(result);
+            try
+            {
+                // Run the build using the published tool
+                var toolAssembly = Path.Combine(publishDir, "rzc.dll");
+                var result = await DotnetMSBuild(
+                    "Build",
+                    $"/p:_RazorForceBuildServer=true /p:_RazorToolAssembly={toolAssembly}",
+                    suppressBuildServer: true); // We don't want to specify a pipe name
 
-            // Shutdown the server
-            var output = new StringWriter();
-            var error = new StringWriter();
-            var application = new Application(CancellationToken.None, Mock.Of<ExtensionAssemblyLoader>(), Mock.Of<ExtensionDependencyChecker>(), (path, properties) => Mock.Of<PortableExecutableReference>(), output, error);
-            var exitCode = application.Execute("shutdown", "-w");
+                Assert.BuildPassed(result);
 
-            Assert.Equal(0, exitCode);
-            Assert.Contains("shut down completed", output.ToString());
+                // Manually shutdown the server
+                var processStartInfo = new ProcessStartInfo()
+                {
+                    WorkingDirectory = publishDir,
+                    UseShellExecute = false,
+                    RedirectStandardError = true,
+                    RedirectStandardOutput = true,
+                    FileName = "dotnet",
+                    Arguments = $"rzc.dll shutdown -w"
+                };
+
+                var logFilePath = Path.Combine(publishDir, "out.log");
+                processStartInfo.Environment.Add("RAZORBUILDSERVER_LOG", logFilePath);
+                var shutdownResult = await MSBuildProcessManager.RunProcessCoreAsync(processStartInfo);
+
+                Assert.Equal(0, shutdownResult.ExitCode);
+                var output = await File.ReadAllTextAsync(logFilePath);
```

Comment:
I couldn't capture the Console logs so using the buildserver log file to capture output.

Reply:
This seems better anyway